topdown/eval.go

```diff
@@ -1682,30 +1685,62 @@ func (e evalBuiltin) eval(iter unifyIterator) error {
 	numDeclArgs := len(e.bi.Decl.FuncArgs().Args)
 
 	e.e.instr.startTimer(evalOpBuiltinCall)
+	var err error
+	needNormalUnify := true
 
-	err := e.f(e.bctx, operands, func(output *ast.Term) error {
-
+	// We skip evaluation of the builtin entirely if the NDBuiltinCache is
+	// present, and we have a non-deterministic builtin already cached.
+	if e.bi.Nondeterministic && e.bctx.NDBuiltinCache != nil {
 		e.e.instr.stopTimer(evalOpBuiltinCall)
 
-		var err error
-
-		if e.bi.Decl.Result() == nil {
-			err = iter()
-		} else if len(operands) == numDeclArgs {
-			if output.Value.Compare(ast.Boolean(false)) != 0 {
-				err = iter()
-			}
-		} else {
-			err = e.e.unify(e.terms[len(e.terms)-1], output, iter)
+		// Unify against the NDBuiltinCache result if present.
+		if v, ok := e.bctx.NDBuiltinCache.Get(ast.NewArray(e.terms...)); ok {
+			err = e.e.unify(e.terms[len(e.terms)-1], v.(*ast.Term), iter)
+			needNormalUnify = false
 		}
 
 		if err != nil {
```

Comment:
Given what happens [below](https://github.com/open-policy-agent/opa/pull/4926/files?diff=split&w=0#diff-997d78d00642bc9f3ec6dde2122442c6a1870122815805212ebdd33de36ea396R1746-R1751) with unwrapping `Halt` errors, can't we do this here,
```go
if err != nil {
	e.e.instr.stopTimer(evalOpBuiltinCall)
	return err
}
```
...and add some `return nil` to avoid the need for `needNormalUnify`? 

I'm sorry, this is just a sketch, but I'm suspecting there some potential for simplification here -- after all, if the lookup doesn't hit something, `err` will be `nil` anyways 🤔

Reply:
It may amuse you to know that this *was* the simplified version from what I initially came up with! :sweat_smile: That said, I am 100% on board with further simplification; the simpler we can make this new evaluator logic, the better.

Reply:
I was able to plumb in a nice early return around the cache lookup, and remove the `needNormalUnify` logic as a result. Thanks for that suggestion! :+1: